nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+
+===Read Methods===
+
+====getProvider====
+Returns information about the  dAPI provider, including who this provider is, the version of their dAPI, and the NEP that the interface is compatible with.
+
+
+'''Method Interface'''
+
+<pre>
+function getProvider(): Promise<Provider>
+</pre>
+
+
+'''Input arguments'''
+
+None
+
+
+'''Success return value'''
+
+object - Provider information defined by the interface below:
+<pre>
+interface Provider {
+  name: string;
+  website: string;
+  version: string;
+  compatibility: string[];
+  extra: object; //this object can contain any attributes specific to the dapi provider, such as an app theme.
+}
+</pre>
+
+The compatibility field will return a list of strings that corresponds to NEPs that the dAPI conforms to. For example:
+<pre>
+compatibility: [
+    'NEP-14',
+    'NEP-23',
+    'NEP-29'
+]
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_DENIED`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+getProvider()
+.then((provider: Provider) => {
+  const {
+    name,
+    website,
+    version,
+    compatibility,
+    extra,
+  } = provider;
+
+  console.log('Provider name: ' + name);
+  console.log('Provider website: ' + website);
+  console.log('Provider dAPI version: ' + version);
+  console.log('Provider dAPI compatibility: ' + JSON.stringify(compatibility));
+  console.log('Extra provider specific atributes: ' + JSON.stringify(compatibility));
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp.');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request, return:
+* The name of the provider wallet
+* The provider website
+* The provider specific version of the wallet
+* A list of NEP for which the specific provider protocol is compatible with
+
+Example
+<pre>
+{
+  name: 'Awesome Wallet',
+  website: 'https://www.awesome.com',
+  version: 'v0.0.1',
+  compatibility: [
+    'NEP-14',
+    'NEP-23',
+    'NEP-29'
+  ],
+  extra: {
+    theme: 'Dark Mode'
+  }
+}
+</pre>
+
+
+====getNetworks====
+Returns the networks the wallet provider has available to connect to, along with the default network the wallet is currently set to.
+
+
+'''Method Interface'''
+
+<pre>
+function getNetworks(): Promise<Networks>
+</pre>
+
+
+'''Input arguments'''
+
+None
+
+
+'''Success return value'''
+
+<pre>
+interface Networks {
+  networks: string[]; // Array of network names the wallet provider has available for the dapp developer to connect to.
```

Comment:
I think that every `network` should itself be an object, carrying two informations:
- name (string)
- protocol (string)

This is important, because a wallet provider can be simultaneously connected to:
- MainNet (Neo2)
- MainNet (Ont)
- MainNet (Neo3)
etc...

Reply:
This allows straightforward implementation of cross-chain stuff and beautiful atomic swaps.

Reply:
Is there ever a case where a user will be connecting to MainNet (Neo2) and TestNet (Ont or Neo3)? I don't believe any of the current wallet providers allow for this.

Reply:
I believe soon we will have use cases for: Neo2 testnet vs neo3 testnet (atomic swaps) and migration, and for ont it should follow I guess.
MainNet vs TestNet I dont see application too, but yet, multiconnection is important (I believe @vncoelho had an issue with Ledger devices not connecting to two networks simultaneously... right?)

Reply:
I think this method should return the `magic number` of the networks. And also, maybe these data can be returned by `getProvider()`.